### ISSUE TITLE
coreclr doesn't build on linux 4.6 and newer

### ISSUE THREAD

**Issue by tmds** [dotnet/coreclr#7345]
OmniSharp is crashing on my freshly installed fedora 23 system. Investigating the issue suggest the crash is somewhere more 'lowlevel' in the stack. See https://github.com/OmniSharp/omnisharp-vscode/issues/773.
The issue might not easily reproduce on a VM. On my native install the crash happens soon after I start using intellisense in vscode. Two coredumps show a segmentation fault below `Thread::RareDisablePreemptiveGC`.

**Comment by tmds** [comment]
I just tried to build the coreclr repo on my fedora box. It is also segfaulting.

when generating version.cpp
```
Welcome to .NET Core!
---------------------
Learn more about .NET Core @ https://aka.ms/dotnet-docs. Use dotnet --help to see available commands or go to https://aka.ms/dotnet-cli-docs.

Telemetry
--------------
The .NET Core tools collect usage data in order to improve your experience. The data is anonymous and does not include commandline arguments. The data is collected by Microsoft and shared with the community.
You can opt out of telemetry by setting a DOTNET_CLI_TELEMETRY_OPTOUT environment variable to 1 using your favorite shell.
You can read more about .NET Core tools telemetry @ https://aka.ms/dotnet-cli-telemetry.

Configuring...
-------------------
A command is running to initially populate your local package cache, to improve restore speed and enable offline access. This command will take up to a minute to complete and will only happen once.
Decompressing 100% 2738 ms
Segmentation fault (core dumped)
```

with skipgenerateversion, later when building the managed components

```
Running: /home/tmds/repos/coreclr/Tools/dotnetcli/dotnet /home/tmds/repos/coreclr/Tools/run.exe /home/tmds/repos/coreclr/config.json build -Project=/home/tmds/repos/coreclr/build.proj -MsBuildLog=/flp:Verbosity=normal;LogFile=/home/tmds/repos/coreclr/bin/Logs/System.Private.CoreLib_Debug.log -BuildTarget -__IntermediatesDir=/home/tmds/repos/coreclr/bin/obj/Linux.x64.Debug -__RootBinDir=/home/tmds/repos/coreclr/bin -BuildNugetPackage=false -UseSharedCompilation=false -BuildArch=x64 -BuildType=Debug -BuildOS=Linux

Welcome to .NET Core!
---------------------
Learn more about .NET Core @ https://aka.ms/dotnet-docs. Use dotnet --help to see available commands or go to https://aka.ms/dotnet-cli-docs.

Telemetry
--------------
The .NET Core tools collect usage data in order to improve your experience. The data is anonymous and does not include commandline arguments. The data is collected by Microsoft and shared with the community.
You can opt out of telemetry by setting a DOTNET_CLI_TELEMETRY_OPTOUT environment variable to 1 using your favorite shell.
You can read more about .NET Core tools telemetry @ https://aka.ms/dotnet-cli-telemetry.

Configuring...
-------------------
A command is running to initially populate your local package cache, to improve restore speed and enable offline access. This command will take up to a minute to complete and will only happen once.
Decompressing 100% 2655 ms
/home/tmds/repos/coreclr/run.sh: line 12: 10303 Segmentation fault      (core dumped) $dotnet $toolRuntime/run.exe $working_tree_root/config.json $*
ERROR: An error occured in /home/tmds/repos/coreclr/Tools/dotnetcli/dotnet /home/tmds/repos/coreclr/Tools/run 11. Check 11 logs under /home/tmds/repos/coreclr.
Failed to build managed components.
```

**Comment by janvorli** [comment]
@tmds I have all of a sudden started to hit a segfault on one of my VMs with Ubuntu 14.04 too. So I went to investigate the issue, but after several attempt where I was hitting the issue running under the lldb, the restore passed and now issue went away and I cannot repro it anymore even after deleting Tools and packages folders. I will try to repro it some way again.
The managed call stack at the failure was as follows:
```
(lldb) clrstack
OS Thread Id: 0x64f7 (10)
        Child SP               IP Call Site
00007FFFEE988B50 00007ffff79c6f2c [HelperMethodFrame_1OBJ: 00007fffee988b50] System.Threading.Monitor.ObjPulse(System.Object)
00007FFFEE988C80 00007FFF7C8B1EED System.Threading.SemaphoreSlim.Release(Int32)
00007FFFEE988CE0 00007FFF7CD3532C System.IO.UnixFileStream.Read(Byte[], Int32, Int32)
00007FFFEE988D10 00007FFF7CD352D8 System.IO.UnixFileStream.Read(Byte[], Int32, Int32)
00007FFFEE988D60 00007FFF7C8EF656 System.IO.BinaryReader.FillBuffer(Int32)
00007FFFEE988D90 00007FFF7C8EE7ED System.IO.BinaryReader.ReadUInt16()
00007FFFEE988DB0 00007FFF7D26A01E System.IO.Compression.ZipCentralDirectoryFileHeader.TryReadBlock(System.IO.BinaryReader, Boolean, System.IO.Compression.ZipCentralDirectoryFileHeader ByRef)
00007FFFEE988E80 00007FFF7D26A5C2 System.IO.Compression.ZipArchive.ReadCentralDirectory()
00007FFFEE988FB0 00007FFF7D26AA8A System.IO.Compression.ZipArchive.get_Entries()
00007FFFEE988FD0 00007FFF7D26ABB3 NuGet.Packaging.ZipArchiveExtensions.GetFiles(System.IO.Compression.ZipArchive)
00007FFFEE989000 00007FFF7D26AD5B NuGet.Packaging.PackageReaderExtensions.GetNuspecFile(NuGet.Packaging.Core.IPackageCoreReader)
00007FFFEE989030 00007FFF7D2F3A18 NuGet.Packaging.PackageExtractor+<>c__DisplayClass2_0+<<InstallFromSourceAsync>b__0>d.MoveNext()
00007FFFEE9891C0 00007FFF7C829C75 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
00007FFFEE989210 00007FFF7C8E90A2 System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
00007FFFEE989250 00007FFF7C8CAA32 System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
00007FFFEE989280 00007FFF7C898DC5 System.Threading.Tasks.Task.FinishContinuations()
00007FFFEE989300 00007FFF7C8A253F System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].TrySetResult(System.Threading.Tasks.VoidTaskResult)
00007FFFEE989320 00007FFF7C8B0F8B System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.VoidTaskResult)
00007FFFEE989360 00007FFF7C8B0EF9 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.Task`1<System.Threading.Tasks.VoidTaskResult>)
00007FFFEE989380 00007FFF7D2F21C0 NuGet.DependencyResolver.SourceRepositoryDependencyProvider+<CopyToAsync>d__16.MoveNext()
00007FFFEE9893F0 00007FFF7C829C75 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
00007FFFEE989440 00007FFF7C8E90A2 System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
00007FFFEE989480 00007FFF7C8CAA32 System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
00007FFFEE9894B0 00007FFF7C898DC5 System.Threading.Tasks.Task.FinishContinuations()
00007FFFEE989530 00007FFF7C8A253F System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].TrySetResult(System.Threading.Tasks.VoidTaskResult)
00007FFFEE989550 00007FFF7C8B0F8B System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.VoidTaskResult)
00007FFFEE989590 00007FFF7C8B0EF9 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]].SetResult(System.Threading.Tasks.Task`1<System.Threading.Tasks.VoidTaskResult>)
00007FFFEE9895B0 00007FFF7C917CBF System.IO.Stream+<CopyToAsyncInternal>d__27.MoveNext()
00007FFFEE989660 00007FFF7C829C75 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
00007FFFEE9896B0 00007FFF7C8E90A2 System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
00007FFFEE9896F0 00007FFF7C8CAA32 System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action, Boolean, System.Threading.Tasks.Task ByRef)
00007FFFEE989720 00007FFF7C898DC5 System.Threading.Tasks.Task.FinishContinuations()
00007FFFEE9897A0 00007FFF7C896F58 System.Threading.Tasks.Task.Finish(Boolean)
00007FFFEE9897E0 00007FFF7C897B00 System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef)
00007FFFEE989890 00007FFF7C897848 System.Threading.Tasks.Task.ExecuteEntry(Boolean)
00007FFFEE9898B0 00007FFF7C8A9EE0 System.Threading.ThreadPoolWorkQueue.Dispatch()
00007FFFEE989C40 00007ffff60b32e7 [DebuggerU2MCatchHandlerFrame: 00007fffee989c40]
```

**Comment by janvorli** [comment]
This repro was with the older tools 1.0.0-rc3-002733. I have synced my VM to the latest sources, now the tools used are 1.0.0-preview3-003223, but I can still repro the crash. The call stack is different now though and seems to happen at random places.
But that made me realize that I have Linux kernel 4.6.2 on my testing VM and we had a bug in coreclr that caused crashes on kernels >= 4.6 (https://github.com/dotnet/coreclr/issues/6016)
The fix was done in https://github.com/dotnet/coreclr/pull/6027 on June 28, 
@tmds can you please check your kernel version (uname -a)?
It looks like the official tools 1.0.0-preview3-003223 don't have this issue fixed.
CC: @mellinoe, @gkhanna79

**Comment by tmds** [comment]
I am using a 4.7.3 kernel.

**Comment by gkhanna79** [comment]
@livarcocc When does CLI pickup LTS of SharedFW?

**Comment by tmds** [comment]
@janvorli behavior matches the referenced issue, after setting COMPlus_INTERNAL_ThreadSuspendInjection to zero dotnet is no longer crashing

**Comment by tmds** [comment]
@livarcocc 
> @livarcocc When does CLI pickup LTS of SharedFW?

**Comment by DustinCampbell** [comment]
@livarcocc -- could chime in on this issue? It's blocking customers using the C# extension for VS Code with an updated kernel.

**Comment by livarcocc** [comment]
@gkhanna79 The CLI has picked up the official 1.0.1 of the shared framework almost a month ago in this PR: https://github.com/dotnet/cli/commit/8d7dd3992fcf943a774d7f96d991b4cbf8470459.

We picked up M.NC.App 1.0.1 and we updated the host to 1.0.4.

**Comment by gkhanna79** [comment]
What is the build of CLI in question that results in the crash? As @janvorli called out, the build he was working with didnt have the fix. It will be good to understand the versions of the involved components as a starting point.

**Comment by tmds** [comment]
My coreclr master branch build has the same version as mentioned by @janvorli
```
/home/tmds/repos/coreclr/Tools/dotnetcli/dotnet --version
1.0.0-preview3-003223
```
The OmniSharp vscode plugin has this version
```
Microsoft .NET Core Shared Framework Host

  Version  : 1.0.1
  Build    : cee57bf6c981237d80aa1631cfe83cb9ba329f12
```
Both stop crashing when COMPlus_INTERNAL_ThreadSuspendInjection is set to zero.

**Comment by DustinCampbell** [comment]
cc @piotrpMSFT as well

**Comment by gkhanna79** [comment]
The commit hash mentioned above maps to https://github.com/dotnet/core-setup/commit/cee57bf6c981237d80aa1631cfe83cb9ba329f12, which is pre-RTM timeframe (and thus, likely corresponds to the RTM build) and thus, definitely does not correspond to https://github.com/dotnet/core-setup/releases/tag/v1.0.1 when we snapped the build containing the fix for Linux kernel.

@tmds If you set COREHOST_TRACE environment variable to 1, you will get a trace log of where binaries are being loaded from. Can you please share the log and also share the build information of libcoreclr.so?

**Comment by tmds** [comment]
@gkhanna79 how can I retrieve the build information of libcoreclr.so?

COREHOST_TRACE:
coreclr Tools: https://gist.githubusercontent.com/tmds/f12cc75b82c5ca52c518852b240b637f/raw/f01b94c407c865f5aaa3cf24baf94f6710871d40/coreclr_Tools
OmniSharp: https://gist.githubusercontent.com/tmds/f12cc75b82c5ca52c518852b240b637f/raw/f01b94c407c865f5aaa3cf24baf94f6710871d40/OmniSharp

**Comment by janvorli** [comment]
@tmds - you can run 
```
strings libcoreclr.so | grep @\(#\)
```
That should get an information in a format like this:
```
@(#)Version 1.0.24410.01 built by: root-89e968e9eaca. Commit Hash: 51131a5a2af24783044c43a60f1052fc35196076
```

**Comment by tmds** [comment]
coreclr Tools and OmniSharp have the same version for libcoreclr.so:
```
@(#)Version 1.0.24214.01. Commit Hash: abbb8f685929c7aeaa087dae46fedc1bc2af4b17
```

**Comment by gkhanna79** [comment]
@tmds According to the trace, libcoreclr.so is being loaded from the following location:

`CoreCLR path = '/home/tmds/repos/coreclr/Tools/dotnetcli/shared/Microsoft.NETCore.App/1.0.0/libcoreclr.so', CoreCLR dir = '/home/tmds/repos/coreclr/Tools/dotnetcli/shared/Microsoft.NETCore.App/1.0.0'`

Do you have a Microsoft.NETCore.App/1.0.1 folder, next to the 1.0.0 above?

Also, please share the trace for "dotnet <path to your published app.dll>".

**Comment by tmds** [comment]
There is only a 1.0.0 folder, no 1.0.1 folder.

COREHOST_TRACE for dotnet: https://gist.githubusercontent.com/tmds/f12cc75b82c5ca52c518852b240b637f/raw/5bba16868d399dce133827d829a90ba391ecb8c5/dotnet

**Comment by tmds** [comment]
So OmniSharp and coreclr Tools are crashing because they are using 1.0.0. When they use 1.0.1 they shouldn't crash anymore.
For OmniSharp, the fix is to target 1.0.1?

my dotnet install (from https://download.microsoft.com/download/1/5/2/1523EBE1-3764-4328-8961-D1BD8ECA9295/dotnet-dev-fedora.23-x64.1.0.0-preview2-003131.tar.gz)
includes 'Microsoft.NETCore.App/1.0.1/'

the coreclr Tools install (from https://dotnetcli.blob.core.windows.net/dotnet/Sdk/1.0.0-preview3-003223/dotnet-dev-fedora.23-x64.1.0.0-preview3-003223.tar.gz)
includes 'Microsoft.NETCore.App/1.0.0'

**Comment by gkhanna79** [comment]
The second trace shows libcoreclr.so is being loaded from 1.0.1 folder:

`CoreCLR path = '/opt/dotnet/shared/Microsoft.NETCore.App/1.0.1/libcoreclr.so', CoreCLR dir = '/opt/dotnet/shared/Microsoft.NETCore.App/1.0.1'`

> When they use 1.0.1 they shouldn't crash anymore.

This is correct.

Since the installer for .NET Core SDK (aka CLI) version 3223 is installing 1.0.0, I suspect you are running into the issue fixed in 1.0.1. This is inline with observation made by @janvorli in https://github.com/dotnet/coreclr/issues/7345#issuecomment-249576324 that 3223 does not have the fix.

@livarcocc PTAL to get this fixed in the upcoming Preview2 release of CLI.

**Comment by gkhanna79** [comment]
CC @leecow @piotrpMSFT

**Comment by DustinCampbell** [comment]
It's a little hard to follow the discussion, but it sounds like this issue *is* fixed but newer versions of the Preview3 versions of the .NET Core SDK don't install it?

OmniSharp has been updated (https://github.com/OmniSharp/omnisharp-roslyn/commit/a607e7df18e8ecd20129d5bad5d40e1091e435f1) but the C# extension for VS Code hasn't been updated to take a new version of it yet.

**Comment by tmds** [comment]
> @livarcocc PTAL to get this fixed in the upcoming Preview2 release of CLI.

It looks like preview2(-003131) has the correct version and preview3(-003223) is using the older version.

**Comment by DustinCampbell** [comment]
@leecow: note that it appears that preview3 still exhibits the issue.

**Comment by piotrpMSFT** [comment]
preview3 003223 is quite an old build. Latest preview3 build in CLI repo is 3748. Perhaps you can try installing latest from https://github.com/dotnet/cli? I'm also curious if we are advertising preview3 installers on any properties. How did preview3 get into this conversation?

**Comment by tmds** [comment]
@piotrpMSFT this is the version @janvorli and I got when building the coreclr in the Tools folder

**Comment by piotrpMSFT** [comment]
Can you try bumping the dependency to 3748 or later?

**Comment by tmds** [comment]
I changed `DotnetCLIVersion.txt` to `1.0.0-preview3-003748`. dotnet is no longer crashing!

coreclr doesn't build.

Made a change to init-tools.sh: `Microsoft.DotNet.BuildTools` -> `microsoft.dotnet.buildtools`.
Now the native part builds.

Managed components fail:
```
Running: /home/tmds/repos/coreclr/Tools/dotnetcli/dotnet /home/tmds/repos/coreclr/Tools/run.exe /home/tmds/repos/coreclr/config.json build -Project=/home/tmds/repos/coreclr/build.proj -MsBuildLog=/flp:Verbosity=normal;LogFile=/home/tmds/repos/coreclr/bin/Logs/System.Private.CoreLib_Debug.log -BuildTarget -__IntermediatesDir=/home/tmds/repos/coreclr/bin/obj/Linux.x64.Debug -__RootBinDir=/home/tmds/repos/coreclr/bin -BuildNugetPackage=false -UseSharedCompilation=false -BuildArch=x64 -BuildType=Debug -BuildOS=Linux
Running: /home/tmds/repos/coreclr/Tools/msbuild.sh /nologo /verbosity:minimal /clp:Summary /maxcpucount  /home/tmds/repos/coreclr/build.proj /p:__BuildType=Debug /p:__BuildArch=x64 /p:__BuildOS=Linux /p:__RootBinDir=/home/tmds/repos/coreclr/bin /p:__IntermediatesDir=/home/tmds/repos/coreclr/bin/obj/Linux.x64.Debug  /flp:Verbosity=normal;LogFile=/home/tmds/repos/coreclr/bin/Logs/System.Private.CoreLib_Debug.log     /p:BuildNugetPackage=false /t:Build /p:UseSharedCompilation=false
/home/tmds/repos/coreclr/Tools/FrameworkTargeting.targets(67,3): error MSB4019: The imported project "/home/tmds/repos/coreclr/Tools/Extensions/Microsoft/Portable/v4.5/Microsoft.Portable.CSharp.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [/home/tmds/repos/coreclr/src/ToolBox/SOS/NETCore/SOS.NETCore.csproj]
```
There is a `v4.5/Microsoft.Portable.CSharp.targets` at `/Tools/microsoft.portable.targets/0.1.1-dev/contentFiles/any/any/Extensions/Microsoft/Portable/v4.5/Microsoft.Portable.CSharp.targets`.

**Comment by janvorli** [comment]
@tmds Can you try to delete the whole Tools and packages directories and retry? Maybe there is something stale in there.

**Comment by tmds** [comment]
@janvorli I cleaned up my working copy before the build. If you give it a try, you should get similar build errors.

**Comment by janvorli** [comment]
@tmds I can confirm this version of tools doesn't work for me either on my Ubuntu 14.04. Actually, early at the beginning, I am getting this error:
```

/home/janvorli/git/coreclr/init-tools.sh: line 137: /home/janvorli/git/coreclr/packages/Microsoft.DotNet.BuildTools/1.0.26-prerelease-00809-01/lib/init-tools.sh: No such file or directory
Running: /home/janvorli/git/coreclr/Tools/dotnetcli/dotnet /home/janvorli/git/coreclr/Tools/run.exe /home/janvorli/git/coreclr/config.json build -Project=/home/janvorli/git/coreclr/build.proj -generateHeaderUnix -NativeVersionSourceFile=/home/janvorli/git/coreclr/bin/obj/Linux.x64.Debug/version.cpp -BuildArch=x64 -BuildType=Debug -BuildOS=Linux
No executable found matching command "dotnet-/home/janvorli/git/coreclr/Tools/run.exe"
ERROR: An error occured in /home/janvorli/git/coreclr/Tools/dotnetcli/dotnet /home/janvorli/git/coreclr/Tools/run 7. Check 7 logs under /home/janvorli/git/coreclr.
```
But the native build starts running after that, so I wonder if you have the same and just haven't noticed this first error that I guess causes the 2nd.

**Comment by janvorli** [comment]
@piotrpMSFT - I have just tried the 1.0.0-preview3-003748 and the init-tools.sh fail with that. I have tried the latest 1.0.0-preview3-003786 and it fails too.
For the latest one, there is a casing issue:
```
/home/janvorli/git/coreclr/init-tools.sh: line 137: /home/janvorli/git/coreclr/packages/Microsoft.DotNet.BuildTools/1.0.26-prerelease-00809-01/lib/init-tools.sh: No such file or directory
```
But `ls /home/janvorli/git/coreclr/packages` shows that there is a folder like that, only with different casing.
```
microsoft.dotnet.buildtools
```

**Comment by gkhanna79** [comment]
@tmds Are you still running into the inability to build? If so, which branch in CoreCLR repo? @mellinoe has been on point to ensure that all supported Platforms are building clean and can help with the issue if it still persists.

**Comment by tmds** [comment]
DotnetCLIVersion.txt on master is still pointing to a dotnet version that
crashes on linux >= 4.6.

Op ma 31 okt. 2016 om 18:38 schreef Gaurav Khanna <notifications@github.com

**Comment by gkhanna79** [comment]
@livarcocc @mellinoe What version should be used that contains the required fix?

**Comment by piotrpMSFT** [comment]
@gkhanna79 @tmds what version of coreclr added the fix? I'll see when that got incorporated.

**Comment by mellinoe** [comment]
Can we just update everything to use the stable CLI coming out of the preview2.1 branch? It's running on the 1.1 RTM shared framework and works on all the distros we care about. Except for Alpine, but we will need to copy that bootstrap version over regardless of what version we choose.

**Comment by piotrpMSFT** [comment]
I think that's the right call @mellinoe. The latest from preview3 dropped PJ support which would likely not play well in coreclr repo at the moment.

**Comment by gkhanna79** [comment]
@piotrpMSFT @livarcocc What are the next steps here?

**Comment by gkhanna79** [comment]
@piotrpMSFT @livarcocc ?

**Comment by livarcocc** [comment]
If I read this right, huge thread, this seems to have been a problem on the preview2-1 CLI, which we don't support anymore. When we shipped 1.0 CLI, it was already depending on the latest runtime, and current CLI out there, 1.0.4, depends on 1.0.5 and 1.1.2 of the CLI. And 2.0 CLI depends on 2.0 runtime. What else is there to be done by the CLI here?

**Comment by tmds** [comment]
@livarcocc I assume this is all fixed by now.
